### ISSUE TITLE
add self-gravity with AMR subcycling-in-time

### ISSUE THREAD

**Issue by BenWibking** [quokka-astro/quokka#334]
This is nontrivial to do in a way that minimizes the inconsistency at the coarse-fine level boundaries.

In order of increasing consistency (and complexity), there are three possible approaches one could take to this:
1. After doing the hydro update on level $l$, perform a Poisson solve on level $l$ only, and add the gravitational acceleration in operator-split fashion. This requires only 1 Poisson solve per level advance. *This is the approach Enzo adopts.*
2. Do step 1, and additionally, after doing a conservative reflux operation, perform a "sync solve" using levels $l$ and $l-1$ to account for mass that has been advected over the coarse-fine boundary during the level $l$ subcycles. Assuming a 2:1 subcycle ratio, this requires 1.5 Poisson solves per level advance (0.5 is due to 1 sync solve per 2 subcycles). *This is the approach ORION2 adopts.*
3. Do steps 1 and 2, except in step 1, correct the solution of the level $l$ Poisson solve so that in the case of no evolution across the level $l$ hydro step, it exactly matches a multi-level Poisson solve performed prior to calling the level $l$ hydro advance. This correction is nonzero because a level $l$ solve does not account for the better resolved mass distribution that is used when doing a composite solve that includes levels $l' > l$. Note that **this method requires 3.5 Poisson solves per level advance** (1 composite solve before the advance, one level solve on level $l$ before the advance, and one level solve on level $l$ after the advance, plus one sync solve for every 2 subcycles). This is expensive, but also allows Strang splitting at no extra cost. *This is the approach Castro and Nyx adopt.*

Regardless of which approach above is adopted in Quokka, the level 0 solve must be handled as a special case, because it has to solve for the Dirichlet boundary conditions that match the free-space/vacuum boundaries we are interested in.

Regardless of the approach above, for level solves at $l > 0$, the Dirichlet boundary conditions are obtained by interpolating the solution previously computed from level $l-1$.

A complete description of option 3 is available in the Castro hydro+gravity paper: https://arxiv.org/pdf/1005.0114.pdf

**Comment by BenWibking** [comment]
@markkrumholz When you have time, let me know what you think. I've added this because it be useful to keep this in mind when thinking about the approach we take to handling particles.

**Comment by BenWibking** [comment]
@pakshingli Let me know what you think about the various approaches for self-gravity I've outlined above. Did you implement self-gravity for ORION2? Is my description of option 2 correct?

**Comment by pakshingli** [comment]
@BenWibking The self-gravity code in ORION2 is coming from the code by Miniati (Miniati & Colella 2007) written using Chombo.  I took the responsibility of the code migration to ORION2.  Your description of the option 2 is correct.  It is described in Miniati & Colella 2007.  Let me read the Nyx paper to better understand the Nyx methods.

**Comment by BenWibking** [comment]
Ok. I have added the number of Poisson solves required for each method in the description above. I think I counted correctly, but let me know if that seems incorrect.

**Comment by pakshingli** [comment]
@BenWibking That is about right.

**Comment by markkrumholz** [comment]
OK, so having read the CASTRO paper, it is honestly not obvious to me whether 2 or 3 is better. I do think we want to do the sync solve, because not doing so causes problems, as shown by [Argertz et al. (2007)](http://adsabs.harvard.edu/abs/2007MNRAS.380..963A) -- specifically, not doing sync solves causes artifacts to appear when self-gravitating structures advect across the grid. I repeated the test described in this paper with ORION, and it performed fine, because there are sync solves. The main argument in the CASTRO paper for 3 is that the extra cost is negligible, because the extra level solves make the multi-level solve converge much faster, since the guess is closer. It is not obvious to me that this is going to be true on GPU, where the overhead of initiating the solve is much higher relative to the cost of iterating. So my suggestion would be to start by implementing 2, and then consider 3 as a future enhancement.

**Comment by BenWibking** [comment]
> I do think we want to do the sync solve, because not doing so causes problems, as shown by [Argertz et al. (2007)](http://adsabs.harvard.edu/abs/2007MNRAS.380..963A) -- specifically, not doing sync solves causes artifacts to appear when self-gravitating structures advect across the grid.

I don't see any simulations with gravity in this paper.

> It is not obvious to me that this is going to be true on GPU, where the overhead of initiating the solve is much higher relative to the cost of iterating. So my suggestion would be to start by implementing 2, and then consider 3 as a future enhancement.

Ok, agreed.

**Comment by markkrumholz** [comment]
Sorry, wrong paper -- and now that I look, maybe Oscar never published the result, and it was just a poster. The test was as follows: initialize a stable Bonnor-Ebert sphere with a uniform, low-density background around it so that the system is in pressure equilibrium. Refine on density, so the BE sphere is refined, while the background is left unrefined. The system should just sit there, and if you give both the BE sphere and the background a uniform velocity, it should just advect peacefully across the grid. However, Oscar demonstrated that in enzo the sphere would be fine if it were not advecting, but if it were advecting then it would spontaneously accelerate or decelarate. This was traced back to the lack of a sync solve: when the sphere advected, if you didn't do the sync, then the potential was not consistent between the coarse and fine levels, and as the sphere crossed from refined to previously-unrefined regions of the volume, it would start to exert self-forces that would cause unphysical acceleration. He thought it was a fundamental issue with AMR, but then I tested it in ORION and showed that it the effect did not occur. That is perhaps why he never published it. It wasn't a general AMR issue, it was an enzo issue. In any event, that's why I think a sync solve is needed.

**Comment by BenWibking** [comment]
I see, that is a useful test. We can certainly do that. In fact, we could set up the problem in Castro and disable either the composite solve or sync solve (or both) and see what happens.

**Comment by psharda** [comment]
@BenWibking any updates here? Would be nice to get back to this now :)

**Comment by BenWibking** [comment]
Yes, this would be nice to have. Unfortunately, it's a lot of work and I don't need it for my current science, so I don't have the time at the moment. Maybe Mark or PS would be able to implement it?

**Comment by pakshingli** [comment]
I am sorry that I am working on implementing sink particles for Quokka, no time on gravity.

**Comment by markkrumholz** [comment]
I know how to do it and could guide someone through it, but realistically I am not going to find time in the immediate future given other commitments.

**Comment by BenWibking** [comment]
@claude Create a PR to address this issue, adopting Approach 1 (the approach Enzo uses), where the Poisson equation is solved only one per level, and there are no other Poisson solves. Create a `PoissonGravity` class that orchestrates the Poisson solve for each level, and handles interpolating the Dirichlet boundary conditions in space and time from the next-coarsest level.
